File: oc-bridge-service/src/main/java/service/RequestManager.java
package service;

import io.netty.handler.codec.http.HttpResponseStatus;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpServerResponse;
import io.vertx.core.json.Json;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.mongo.MongoClient;
import io.vertx.ext.web.RoutingContext;
import org.yaml.snakeyaml.Yaml;

import java.io.InputStream;
import java.util.Map;
import java.util.UUID;

class RequestManager {
    private static Vertx vertx;
    private static JsonObject config = new JsonObject();
    private static final String DB_CONNECTION_STRING = "connection_string";
    private static final String DOCUMENT_ID = "_id";
    private static final String EVENT_ID = "eventId";
    private static final String COLLECTION_NAME = "events";
    private static final String DUPLICATED_KEY_CODE = "E11000";


    static void initializeRequestManager(Vertx vertx) {
        RequestManager.vertx = vertx;

        Yaml yaml = new Yaml();
        InputStream inputStream = OcBridge.class
                .getClassLoader()
                .getResourceAsStream("mongoConfig.yaml");
        Map<String, Object> obj = yaml.load(inputStream);
        config.put(DB_CONNECTION_STRING, obj.get(DB_CONNECTION_STRING));
    }

    static void handleCreateEvent(RoutingContext routingContext) {
        HttpServerResponse response = routingContext.response();
        String uuid = UUID.randomUUID().toString();
        JsonObject document = new JsonObject().put(DOCUMENT_ID, uuid);

        MongoClient.createNonShared(vertx, config).insert(COLLECTION_NAME, document, result -> {
            if (result.succeeded()) {
                response
                        .putHeader("Content-Type", "application/json")
                        .setStatusCode(HttpResponseStatus.CREATED.code())
                        .end(Json.encodePrettily(document));
            } else if (isDuplicateKey(result.cause().getMessage())) {
                handleCreateEvent(routingContext);
            }
            else {
                response.setStatusCode(HttpResponseStatus.INTERNAL_SERVER_ERROR.code()).end();
            }
        });
    }

    private static boolean isDuplicateKey(String errorMessage) {
        return errorMessage.startsWith(DUPLICATED_KEY_CODE);
    }

    static void handleGetEventById(RoutingContext routingContext){
        HttpServerResponse response = routingContext.response();
        String eventId = routingContext.request().getParam(EVENT_ID);
        try{
            UUID.fromString(eventId);
            JsonObject query = new JsonObject().put(DOCUMENT_ID, eventId);
            MongoClient.createNonShared(vertx, config).find(COLLECTION_NAME, query, result -> {
                if(result.succeeded()){
                    try {
                        JsonObject resultJson = result.result().get(0);
                        resultJson.remove(DOCUMENT_ID);
                        if (resultJson.size() > 0)
                            response
                                    .putHeader("Content-Type", "application/json")
                                    .setStatusCode(HttpResponseStatus.OK.code())
                                    .end(Json.encodePrettily(resultJson));
                        else
                            response.setStatusCode(HttpResponseStatus.NO_CONTENT.code()).end();
                    } catch (IndexOutOfBoundsException ex) {
                        response.setStatusCode(HttpResponseStatus.BAD_REQUEST.code()).end();
                    }
                }else {
                    response.setStatusCode(HttpResponseStatus.INTERNAL_SERVER_ERROR.code()).end();
                }
            });
        } catch (IllegalArgumentException exception){
            response.setStatusCode(HttpResponseStatus.NOT_FOUND.code()).end();
        }
    }

    static void handleUpdateEvent(RoutingContext routingContext){
        HttpServerResponse response = routingContext.response();
        JsonObject body = routingContext.getBodyAsJson();
        JsonObject query = new JsonObject().put(DOCUMENT_ID, routingContext.request().getParam(EVENT_ID));
        JsonObject update = new JsonObject().put("$set", body);
        MongoClient.createNonShared(vertx, config).updateCollection(COLLECTION_NAME, query, update, res -> {
            if (res.succeeded()) {
                if (res.result().getDocModified() == 0)
                    response.setStatusCode(HttpResponseStatus.BAD_REQUEST.code()).end();
                else
                    response.setStatusCode(HttpResponseStatus.NO_CONTENT.code()).end();
            } else {
                response.setStatusCode(HttpResponseStatus.INTERNAL_SERVER_ERROR.code()).end();
            }
        });
    }
}

File: oc-bridge-service/src/main/kotlin/service/RequestManager.kt
package service

import io.netty.handler.codec.http.HttpResponseStatus.BAD_REQUEST
import io.netty.handler.codec.http.HttpResponseStatus.CREATED
import io.netty.handler.codec.http.HttpResponseStatus.INTERNAL_SERVER_ERROR
import io.netty.handler.codec.http.HttpResponseStatus.NOT_FOUND
import io.netty.handler.codec.http.HttpResponseStatus.NO_CONTENT
import io.netty.handler.codec.http.HttpResponseStatus.OK
import io.vertx.core.Vertx
import io.vertx.core.json.Json
import io.vertx.core.json.JsonObject
import io.vertx.ext.mongo.MongoClient
import io.vertx.ext.web.RoutingContext

import java.util.UUID

object RequestManager {
    private var vertx: Vertx? = null
    private val CONFIG = JsonObject().put("connection_string", "mongodb://loveeclipse:PC-preh2019@ds149676.mlab.com:49676/heroku_jw7pjmcr")
    private const val DOCUMENT_ID = "_id"
    private const val EVENT_ID = "eventId"
    private const val COLLECTION_NAME = "events"
    private const val DUPLICATED_KEY_CODE = "E11000"
    private const val SECONDARY = "secondary"

    fun initializeRequestManager(vertx: Vertx) {
        RequestManager.vertx = vertx
    }

    fun createEvent(routingContext: RoutingContext) {
        val response = routingContext.response()
        val uuid = UUID.randomUUID().toString()
        val document = JsonObject().put(DOCUMENT_ID, uuid)

        MongoClient.createNonShared(vertx, CONFIG).insert(COLLECTION_NAME, document) { result ->
            when {
                result.succeeded() -> response
                        .putHeader("Content-Type", "application/json")
                        .setStatusCode(CREATED.code())
                        .end(Json.encodePrettily(document))
                isDuplicateKey(result.cause().message) -> createEvent(routingContext)
                else -> response.setStatusCode(INTERNAL_SERVER_ERROR.code()).end()
            }
        }
    }

    private fun isDuplicateKey(errorMessage: String?) = errorMessage?.startsWith(DUPLICATED_KEY_CODE) ?: false

    fun retrieveEventById(routingContext: RoutingContext) {
        val response = routingContext.response()
        val eventId = routingContext.request().getParam(EVENT_ID)
        try {
            UUID.fromString(eventId)
            val query = JsonObject().put(DOCUMENT_ID, eventId)
            MongoClient.createNonShared(vertx, CONFIG).find(COLLECTION_NAME, query) { result ->
                if (result.succeeded()) {
                    try {
                        val resultJson = result.result()[0]
                        resultJson.remove(DOCUMENT_ID)
                        if (resultJson.size() > 0)
                            response
                                    .putHeader("Content-Type", "application/json")
                                    .setStatusCode(OK.code())
                                    .end(Json.encodePrettily(resultJson))
                        else
                            response.setStatusCode(NO_CONTENT.code()).end()
                    } catch (ex: IndexOutOfBoundsException) {
                        response.setStatusCode(NOT_FOUND.code()).end()
                    }
                } else {
                    response.setStatusCode(INTERNAL_SERVER_ERROR.code()).end()
                }
            }
        } catch (exception: IllegalArgumentException) {
            response.setStatusCode(BAD_REQUEST.code()).end()
        }
    }

    fun updateEvent(routingContext: RoutingContext) {
        val response = routingContext.response()
        val eventId = routingContext.request().getParam(EVENT_ID)
        try {
            UUID.fromString(eventId)
            val body = routingContext.bodyAsJson
            if (body.containsKey(SECONDARY))
                body.getBoolean(SECONDARY)
            val query = JsonObject().put(DOCUMENT_ID, eventId)
            val update = JsonObject().put("\$set", body)
            MongoClient.createNonShared(vertx, CONFIG).updateCollection(COLLECTION_NAME, query, update) { res ->
                if (res.succeeded()) {
                    when {
                        res.result().docModified == 0L -> response.setStatusCode(NOT_FOUND.code()).end()
                        else -> response.setStatusCode(OK.code()).end()
                    }
                } else {
                    response.setStatusCode(INTERNAL_SERVER_ERROR.code()).end()
                }
            }
        } catch (exception: Exception) {
            response.setStatusCode(BAD_REQUEST.code()).end()
        }
    }
}
